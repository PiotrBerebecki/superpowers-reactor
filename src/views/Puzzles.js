import React, { Component } from 'react';

import { Main } from './../styled';

class Puzzles extends Component {
  render() {
    return (
      <Main>
        Puzzles {this.props.category}
      </Main>
    );
  }
}

export default Puzzles;
