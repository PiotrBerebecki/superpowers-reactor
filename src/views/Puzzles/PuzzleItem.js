import React, { Component } from 'react';

import {
  PuzzleItemWrapper,
  PuzzleTitle,
  PuzzleText,
  AnswerButton,
  AnswerText,
} from './../../styled';

class PuzzleItem extends Component {
  state = {
    isAnswerVisible: false,
  };

  toggleAnswer = () => {
    this.setState(prevState => ({
      isAnswerVisible: !prevState.isAnswerVisible,
    }));
  };

  render() {
    return (
      <PuzzleItemWrapper>
        <PuzzleTitle>{this.props.title}</PuzzleTitle>
        <PuzzleText>{this.props.text}</PuzzleText>
        <AnswerButton onClick={this.toggleAnswer}>Show Answer</AnswerButton>
        {this.state.isAnswerVisible &&
          <AnswerText>{this.props.answer}</AnswerText>}
      </PuzzleItemWrapper>
    );
  }
}

export default PuzzleItem;
