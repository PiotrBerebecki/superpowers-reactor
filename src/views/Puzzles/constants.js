export const GET_PUZZLES = 'superpowers/Puzzles/GET_PUZZLES';
export const GET_PUZZLES_SUCCEEDED =
  'superpowers/Puzzles/GET_PUZZLES_SUCCEEDED';
export const GET_PUZZLES_FAILED = 'superpowers/Puzzles/GET_PUZZLES_FAILED';

export const POPULAR = 'popular';
export const LATEST = 'latest';
export const FAVOURITE = 'favourite';
