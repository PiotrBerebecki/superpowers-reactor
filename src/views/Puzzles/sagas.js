import { call, put, takeLatest } from 'redux-saga/effects';

import {
  GET_PUZZLES,
  GET_PUZZLES_SUCCEEDED,
  GET_PUZZLES_FAILED,
  POPULAR,
  LATEST,
  FAVOURITE,
} from './constants';
import preloadedPuzzles from './../../database/puzzles';

export function getPuzzles(category) {
  return {
    type: GET_PUZZLES,
    category,
  };
}

// to be replaced with real API
export function fetchPuzzles(category) {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(preloadedPuzzles);
    }, 0);
  });
}

export function sortPuzzles(puzzles, category) {
  switch (category) {
    case POPULAR:
      return [...puzzles].sort((a, b) => b.favCount - a.favCount);
    case LATEST:
      return [...puzzles].sort((a, b) => b.id - a.id);
    case FAVOURITE:
      return puzzles.filter(puzzle => puzzle.isFavourite);
    default:
      return puzzles;
  }
}

export function* handlePuzzlesSaga(action) {
  try {
    const puzzles = yield call(fetchPuzzles, action.category);

    const sortedPuzzles = sortPuzzles(puzzles, action.category);
    yield put({ type: GET_PUZZLES_SUCCEEDED, payload: sortedPuzzles });
  } catch (e) {
    yield put({
      type: GET_PUZZLES_FAILED,
      payload: [
        {
          title: 'Sorry, there has been a problem fetching the requested information.',
        },
      ],
    });
  }
}

export default function* puzzlesSaga() {
  yield takeLatest(GET_PUZZLES, handlePuzzlesSaga);
}
