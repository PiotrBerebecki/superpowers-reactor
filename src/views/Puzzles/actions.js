import { GET_PUZZLES } from './constants';

export function getPuzzles(category) {
  return {
    type: GET_PUZZLES,
    category,
  };
}
