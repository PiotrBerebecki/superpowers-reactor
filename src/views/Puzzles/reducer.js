import { GET_PUZZLES } from './constants';

function puzzlesReducer(state = [], action) {
  switch (action.type) {
    case GET_PUZZLES:
      return state;
    default:
      return state;
  }
}

export default puzzlesReducer;
