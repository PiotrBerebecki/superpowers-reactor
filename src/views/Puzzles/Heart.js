import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { StyledHeart } from './../../styled/index';
import { toggleFavourite } from './sagas';

function Heart({ isFavourite, id, toggleFavourite }) {
  const handleClick = () => toggleFavourite(id);
  return (
    <StyledHeart
      isFavourite={isFavourite}
      onClick={handleClick}
      viewBox="0 0 32 32"
      preserveAspectRatio="xMidYMid meet"
    >
      <path d="M16.042 8.345c0 0-2-4.262-6.5-4.262-4.917 0-7.5 4.167-7.5 8.333 0 6.917 14 15.5 14 15.5s13.916-8.5 13.916-15.5c0-4.25-2.666-8.333-7.416-8.333s-6.5 4.262-6.5 4.262z" />
    </StyledHeart>
  );
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators({ toggleFavourite }, dispatch);
};

export default connect(null, mapDispatchToProps)(Heart);
