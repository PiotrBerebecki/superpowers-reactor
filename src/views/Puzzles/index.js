import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import uuid from 'uuid/v4';

import { getPuzzles } from './actions';
import Main from './../../app/Main';
import PuzzleItem from './PuzzleItem';

import { MainWrapper, MainContainer, PuzzleList } from './../../styled';

export class Puzzles extends Component {
  getCategory = props => {
    return (
      props.category ||
      (props.match && props.match.path && props.match.path.slice(1))
    );
  };

  componentWillReceiveProps(nextProps) {
    const currentCategory = this.getCategory(this.props);
    const nextCategory = this.getCategory(nextProps);
    if (currentCategory !== nextCategory) {
      this.props.getPuzzles(this.getCategory(nextProps));
    }
  }

  componentDidMount() {
    this.props.getPuzzles(this.getCategory(this.props));
  }

  render() {
    const renderPuzzles = this.props.puzzles.map(puzzle => (
      <PuzzleItem key={uuid()} {...puzzle} />
    ));

    return (
      <Main>
        <PuzzleList>
          {renderPuzzles}
        </PuzzleList>
      </Main>
    );
  }
}

export const mapStateToProps = state => {
  return {
    puzzles: state.puzzles,
  };
};

export const mapDispatchToProps = dispatch => {
  return bindActionCreators({ getPuzzles }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(Puzzles);
