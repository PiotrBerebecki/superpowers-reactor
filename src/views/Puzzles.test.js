/**
 * @jest-environment node
 */

import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { shallowToJson } from 'enzyme-to-json';
import 'jest-styled-components';

import Puzzles from './Puzzles';

describe('Puzzles', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<Puzzles />);
    expect(tree).toMatchStyledComponentsSnapshot();
  });

  test('renders without crashing', () => {
    const wrapper = shallow(<Puzzles />);
    expect(shallowToJson(wrapper)).toMatchStyledComponentsSnapshot();
  });
});
