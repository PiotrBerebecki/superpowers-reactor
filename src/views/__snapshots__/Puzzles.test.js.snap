// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Puzzles renders correctly 1`] = `
.hfOXWO {
  background-color: #ddd;
}

<main
  className="sc-bZQynM hfOXWO"
>
  Puzzles 
</main>
`;

exports[`Puzzles renders without crashing 1`] = `
<styled.main>
  Puzzles 
</styled.main>
`;
